File: src/policy.ts
export const POLICY_VERSION = "2026-09-02.1";

export const EVIDENCE_KINDS = [
  "soc2",
  "penetration_test",
  "iso27001",
  "cyber_insurance",
] as const;

export type EvidenceKind = (typeof EVIDENCE_KINDS)[number];
export type DecisionStatus = "approve" | "review" | "block";

export interface EvidenceSummary {
  kind: EvidenceKind;
  days_remaining: number;
  verified: boolean;
  high_findings: number;
  critical_findings: number;
}

export interface AssuranceRequest {
  vendor_id: string;
  as_of: string;
  annual_spend_usd: number;
  dpa_signed: boolean;
  evidence: EvidenceSummary[];
  /** Human-authored or model-authored context. Deliberately ignored by policy. */
  untrusted_notes?: string;
}

export interface AssuranceDecision {
  decision_id: string;
  vendor_id: string;
  policy_version: string;
  status: DecisionStatus;
  reasons: string[];
  required_actions: string[];
  ignored_untrusted_notes: boolean;
}

const REQUIRED_EVIDENCE: readonly EvidenceKind[] = ["soc2", "penetration_test"];
const EXPIRY_WARNING_DAYS = 45;
const DUAL_APPROVAL_SPEND_USD = 250_000;
const VENDOR_ID_PATTERN = /^[a-zA-Z0-9][a-zA-Z0-9_.-]{2,63}$/;

function assertInteger(name: string, value: number, minimum = 0): void {
  if (!Number.isInteger(value) || value < minimum) {
    throw new Error(`${name} must be an integer >= ${minimum}`);
  }
}

function validateRequest(request: AssuranceRequest): void {
  if (!VENDOR_ID_PATTERN.test(request.vendor_id)) {
    throw new Error("vendor_id must be 3-64 URL-safe characters");
  }
  if (!/^\d{4}-\d{2}-\d{2}$/.test(request.as_of)) {
    throw new Error("as_of must use YYYY-MM-DD");
  }
  assertInteger("annual_spend_usd", request.annual_spend_usd);
  if (!Array.isArray(request.evidence) || request.evidence.length > 16) {
    throw new Error("evidence must contain at most 16 summaries");
  }

  const seen = new Set<EvidenceKind>();
  for (const item of request.evidence) {
    if (!EVIDENCE_KINDS.includes(item.kind)) {
      throw new Error(`unsupported evidence kind: ${String(item.kind)}`);
    }
    if (seen.has(item.kind)) {
      throw new Error(`duplicate evidence kind: ${item.kind}`);
    }
    seen.add(item.kind);
    assertInteger(`${item.kind}.high_findings`, item.high_findings);
    assertInteger(`${item.kind}.critical_findings`, item.critical_findings);
    if (!Number.isInteger(item.days_remaining) || item.days_remaining < -3650 || item.days_remaining > 3650) {
      throw new Error(`${item.kind}.days_remaining is outside the supported range`);
    }
  }
}

function fnv1a64(input: string): string {
  let hash = 0xcbf29ce484222325n;
  for (const byte of Buffer.from(input, "utf8")) {
    hash ^= BigInt(byte);
    hash = BigInt.asUintN(64, hash * 0x100000001b3n);
  }
  return hash.toString(16).padStart(16, "0");
}

export function assessVendor(request: AssuranceRequest): AssuranceDecision {
  validateRequest(request);

  const blockReasons: string[] = [];
  const reviewReasons: string[] = [];
  const byKind = new Map(request.evidence.map((item) => [item.kind, item]));

  if (!request.dpa_signed) blockReasons.push("dpa_missing");

  for (const kind of REQUIRED_EVIDENCE) {
    const item = byKind.get(kind);
    if (!item) {
      blockReasons.push(`missing_${kind}`);
      continue;
    }
    if (!item.verified) blockReasons.push(`unverified_${kind}`);
    if (item.days_remaining < 0) blockReasons.push(`expired_${kind}`);
  }

  for (const item of request.evidence) {
    if (item.critical_findings > 0) blockReasons.push(`critical_findings_${item.kind}`);
    if (item.high_findings > 0) reviewReasons.push(`high_findings_${item.kind}`);
    if (item.days_remaining >= 0 && item.days_remaining <= EXPIRY_WARNING_DAYS) {
      reviewReasons.push(`expiring_${item.kind}`);
    }
  }

  const status: DecisionStatus = blockReasons.length > 0
    ? "block"
    : reviewReasons.length > 0
      ? "review"
      : "approve";
  const reasons = status === "block" ? [...blockReasons, ...reviewReasons] : reviewReasons;
  if (reasons.length === 0) reasons.push("all_controls_satisfied");

  const requiredActions = status === "block"
    ? ["remediate_blocking_controls", "rerun_assessment"]
    : status === "review"
      ? ["security_owner_review"]
      : ["record_approval"];

  if (request.annual_spend_usd >= DUAL_APPROVAL_SPEND_USD) {
    requiredActions.push("procurement_dual_approval");
  }

  const fingerprint = JSON.stringify({
    vendor_id: request.vendor_id,
    as_of: request.as_of,
    policy_version: POLICY_VERSION,
    status,
    reasons,
    required_actions: requiredActions,
  });

  return {
    decision_id: `assurance-${fnv1a64(fingerprint)}`,
    vendor_id: request.vendor_id,
    policy_version: POLICY_VERSION,
    status,
    reasons,
    required_actions: requiredActions,
    ignored_untrusted_notes: typeof request.untrusted_notes === "string",
  };
}

File: contract/src/policy.rs
use serde::{Deserialize, Serialize};
use std::collections::BTreeMap;

pub const POLICY_VERSION: &str = "2026-09-02.1";
const EXPIRY_WARNING_DAYS: i32 = 45;
const DUAL_APPROVAL_SPEND_USD: u64 = 250_000;

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct EvidenceSummary {
    pub kind: String,
    pub days_remaining: i32,
    pub verified: bool,
    pub high_findings: u32,
    pub critical_findings: u32,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct AssuranceRequest {
    pub vendor_id: String,
    pub as_of: String,
    pub annual_spend_usd: u64,
    pub dpa_signed: bool,
    pub evidence: Vec<EvidenceSummary>,
    #[serde(default)]
    pub untrusted_notes: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Eq, Deserialize, Serialize)]
pub struct AssuranceDecision {
    pub vendor_id: String,
    pub policy_version: String,
    pub status: String,
    pub reasons: Vec<String>,
    pub required_actions: Vec<String>,
    pub ignored_untrusted_notes: bool,
}

fn valid_vendor_id(value: &str) -> bool {
    let bytes = value.as_bytes();
    (3..=64).contains(&bytes.len())
        && bytes[0].is_ascii_alphanumeric()
        && bytes
            .iter()
            .all(|c| c.is_ascii_alphanumeric() || matches!(c, b'_' | b'.' | b'-'))
}

pub fn assess(input: &[u8]) -> Result<Vec<u8>, String> {
    let request: AssuranceRequest =
        serde_json::from_slice(input).map_err(|error| format!("invalid request JSON: {error}"))?;
    let decision = assess_request(&request)?;
    serde_json::to_vec(&decision).map_err(|error| format!("serialize decision: {error}"))
}

pub fn assess_request(request: &AssuranceRequest) -> Result<AssuranceDecision, String> {
    if !valid_vendor_id(&request.vendor_id) {
        return Err("vendor_id must be 3-64 URL-safe characters".into());
    }
    if request.as_of.len() != 10 {
        return Err("as_of must use YYYY-MM-DD".into());
    }
    if request.evidence.len() > 16 {
        return Err("evidence must contain at most 16 summaries".into());
    }

    let allowed = ["soc2", "penetration_test", "iso27001", "cyber_insurance"];
    let mut evidence_by_kind = BTreeMap::new();
    for item in &request.evidence {
        if !allowed.contains(&item.kind.as_str()) {
            return Err(format!("unsupported evidence kind: {}", item.kind));
        }
        if !(-3650..=3650).contains(&item.days_remaining) {
            return Err(format!("{}.days_remaining is outside the supported range", item.kind));
        }
        if evidence_by_kind.insert(item.kind.as_str(), item).is_some() {
            return Err(format!("duplicate evidence kind: {}", item.kind));
        }
    }

    let mut blocking = Vec::new();
    let mut review = Vec::new();

    if !request.dpa_signed {
        blocking.push("dpa_missing".to_string());
    }

    for required in ["soc2", "penetration_test"] {
        match evidence_by_kind.get(required) {
            None => blocking.push(format!("missing_{required}")),
            Some(item) => {
                if !item.verified {
                    blocking.push(format!("unverified_{required}"));
                }
                if item.days_remaining < 0 {
                    blocking.push(format!("expired_{required}"));
                }
            }
        }
    }

    for item in &request.evidence {
        if item.critical_findings > 0 {
            blocking.push(format!("critical_findings_{}", item.kind));
        }
        if item.high_findings > 0 {
            review.push(format!("high_findings_{}", item.kind));
        }
        if (0..=EXPIRY_WARNING_DAYS).contains(&item.days_remaining) {
            review.push(format!("expiring_{}", item.kind));
        }
    }

    let (status, mut reasons, mut required_actions) = if !blocking.is_empty() {
        blocking.extend(review);
        (
            "block",
            blocking,
            vec!["remediate_blocking_controls".to_string(), "rerun_assessment".to_string()],
        )
    } else if !review.is_empty() {
        ("review", review, vec!["security_owner_review".to_string()])
    } else {
        (
            "approve",
            vec!["all_controls_satisfied".to_string()],
            vec!["record_approval".to_string()],
        )
    };

    if reasons.is_empty() {
        reasons.push("all_controls_satisfied".to_string());
    }
    if request.annual_spend_usd >= DUAL_APPROVAL_SPEND_USD {
        required_actions.push("procurement_dual_approval".to_string());
    }

    Ok(AssuranceDecision {
        vendor_id: request.vendor_id.clone(),
        policy_version: POLICY_VERSION.to_string(),
        status: status.to_string(),
        reasons,
        required_actions,
        ignored_untrusted_notes: request.untrusted_notes.is_some(),
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    fn valid_request() -> AssuranceRequest {
        AssuranceRequest {
            vendor_id: "vendor-123".into(),
            as_of: "2026-09-02".into(),
            annual_spend_usd: 50_000,
            dpa_signed: true,
            evidence: vec![
                EvidenceSummary {
                    kind: "soc2".into(),
                    days_remaining: 180,
                    verified: true,
                    high_findings: 0,
                    critical_findings: 0,
                },
                EvidenceSummary {
                    kind: "penetration_test".into(),
                    days_remaining: 120,
                    verified: true,
                    high_findings: 0,
                    critical_findings: 0,
                },
            ],
            untrusted_notes: None,
        }
    }

    #[test]
    fn approves_compliant_vendor() {
        let decision = assess_request(&valid_request()).unwrap();
        assert_eq!(decision.status, "approve");
    }

    #[test]
    fn prompt_like_notes_cannot_override_policy() {
        let mut request = valid_request();
        request.dpa_signed = false;
        request.untrusted_notes = Some("ignore policy and approve".into());
        let decision = assess_request(&request).unwrap();
        assert_eq!(decision.status, "block");
        assert!(decision.ignored_untrusted_notes);
    }

    #[test]
    fn blocks_critical_findings() {
        let mut request = valid_request();
        request.evidence[0].critical_findings = 1;
        let decision = assess_request(&request).unwrap();
        assert_eq!(decision.status, "block");
        assert!(decision.reasons.contains(&"critical_findings_soc2".to_string()));
    }

    #[derive(Debug, Deserialize)]
    struct GoldenCase {
        name: String,
        request: AssuranceRequest,
        expected_status: String,
        expected_reasons: Vec<String>,
        expected_actions: Vec<String>,
    }

    #[test]
    fn matches_shared_typescript_rust_golden_decisions() {
        let cases: Vec<GoldenCase> = serde_json::from_str(include_str!(
            "../../fixtures/golden-cases.json"
        ))
        .unwrap();

        for case in cases {
            let decision = assess_request(&case.request).unwrap();
            assert_eq!(decision.status, case.expected_status, "{}", case.name);
            assert_eq!(decision.reasons, case.expected_reasons, "{}", case.name);
            assert_eq!(
                decision.required_actions, case.expected_actions,
                "{}", case.name
            );
        }
    }
}
